pyo3-derive-backend/src/frompy.rs

```diff
@@ -0,0 +1,428 @@
+    Struct(Vec<(&'a Ident, ExprCall)>),
+    /// Newtype struct container, e.g. `#[transparent] struct Foo { a: String }`
+    ///
+    /// The field specified by the identifier is extracted directly from the object.
+    StructNewtype(&'a Ident),
+    /// Tuple struct, e.g. `struct Foo(String)`.
+    ///
+    /// Fields are extracted from a tuple.
+    Tuple(usize),
+    /// Tuple newtype, e.g. `#[transparent] struct Foo(String)`
+    ///
+    /// The wrapped field is directly extracted from the object.
+    TupleNewtype,
+}
+
+/// Data container
+///
+/// Either describes a struct or an enum variant.
+#[derive(Debug)]
+struct Container<'a> {
+    ident: &'a Ident,
+    style: Style<'a>,
+    err_name: String,
+}
+
+impl<'a> Container<'a> {
+    /// Construct a container from an enum Variant.
+    ///
+    /// Fails if the variant has no fields or incompatible attributes.
+    fn from_variant(var: &'a Variant) -> Result<Self> {
+        Self::new(&var.fields, &var.ident, &var.attrs)
+    }
+
+    /// Construct a container based on fields, identifier and attributes.
+    ///
+    /// Fails if the variant has no fields or incompatible attributes.
+    fn new(fields: &'a Fields, ident: &'a Ident, attrs: &'a [Attribute]) -> Result<Self> {
+        let transparent = attrs.iter().any(|a| a.path.is_ident("transparent"));
+        if transparent {
+            Self::check_transparent_len(fields)?;
+        }
+        let style = match fields {
+            Fields::Unnamed(unnamed) => {
+                if transparent {
+                    Style::TupleNewtype
+                } else {
+                    Style::Tuple(unnamed.unnamed.len())
+                }
+            }
+            Fields::Named(named) => {
+                if transparent {
+                    let field = named
+                        .named
+                        .iter()
+                        .next()
+                        .expect("Check for len 1 is done above");
+                    let ident = field
+                        .ident
+                        .as_ref()
+                        .expect("Named fields should have identifiers");
+                    Style::StructNewtype(ident)
+                } else {
+                    let mut fields = Vec::new();
+                    for field in named.named.iter() {
+                        let ident = field
+                            .ident
+                            .as_ref()
+                            .expect("Named fields should have identifiers");
+                        fields.push((ident, ext_fn(&field.attrs, ident)?))
+                    }
+                    Style::Struct(fields)
+                }
+            }
+            Fields::Unit => {
+                return Err(syn::Error::new_spanned(
+                    &fields,
+                    "Cannot derive FromPyObject for Unit structs and variants",
+                ))
```

Comment:
Again would be nice to see what serde does here with json. It might be acceptable for `None` to be provided?